src/main/java/io/nitric/api/document/SubCollection.java

```diff
@@ -0,0 +1,119 @@
+package io.nitric.api.document;
+/*-
+ * #%L
+ * Nitric Java SDK
+ * %%
+ * Copyright (C) 2021 Nitric Technologies Pty Ltd
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import java.util.Map;
+import io.nitric.util.Contracts;
+
+/**
+ * Provides a document Sub Collection class.
+ */
+public class SubCollection extends AbstractCollection {
+
+    // Constructor ------------------------------------------------------------
+
+    /*
+     * Enforce package builder patterns.
+     */
+    SubCollection(DocColl collection) {
+        super(collection);
+
+        Contracts.requireNonNull(collection.parent, "collection.parent");
+    }
+
+    // Public Methods ---------------------------------------------------------
+
+    /**
+     * Create a new collection Document Reference object for given document id and <code>Map</code> value type.
+     *
+     * @param id the document unique id (required)
+     * @return new collection document reference
+     */
+    @Override
+    public SubCollectionDocRef<Map> doc(String id) {
+        Contracts.requireNonBlank(id, "id");
+
+        if (collection.parent.id.isBlank()) {
+            throw newMissingParentException(id);
+        }
+
+        var docKey = new DocKey(collection, id);
+        return new SubCollectionDocRef<Map>(docKey, Map.class);
+    }
+
+    /**
+     * Create a new collection Document Reference object for given document id and value type.
+     *
+     * @param id   the document unique id (required)
+     * @param type the document value type (required)
+     * @return new collection document reference
+     */
+    @Override
+    public <K> SubCollectionDocRef<K> doc(String id, Class<K> type) {
+        Contracts.requireNonBlank(id, "id");
+
+        if (collection.parent.id.isBlank()) {
+            throw newMissingParentException(id);
```

Comment:
Is this just to subcollection queries without specific parent document scope? It seems a bit fragile to implement a full interface but only allow half its functionality under specific circumstances, and the change in behaviour isn't necessarily immediately apparent to a consumer as well.

I think a less fragile design would be to be able to set the scope within a query object itself at a higher level. Would be good to discuss further. In firestore `CollectionGroups` are used, while we might not be able to facilitate this kind of functionality directly I think the general interface they present is more robust.

Reply:
This check is to ensure a subcollection document reference has a valid parent path hierarchy. For example below:

```
var doc = Documents.collection("customers").collection("orders").doc("123")

/collections/<unknown>/orders/123
```